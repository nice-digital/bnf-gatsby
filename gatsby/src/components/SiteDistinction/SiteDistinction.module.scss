@use '../../styles/colours';
@use '../../styles/settings';
@use '@nice-digital/nds-core/scss/focus';
@use '@nice-digital/nds-core/scss/media-queries';
@use '@nice-digital/nds-core/scss/utils';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/typography';

.wrapper {
  border-top: utils.rem(spacing.$x-small) solid;
  position: sticky;
  top: -0.1px; // Hack to avoid 1px gap due to some weird Chrome sub pixel rounding
  z-index: 101; // Has to be visible over the sticky section nav (z: 100) and back to top (z:99)

  &.bnf {
    border-top-color: settings.$bnf-brand-colour;
  }

  &.bnfc {
    border-top-color: settings.$bnfc-brand-colour;
  }
}

.container {
  position: relative;
}

.tabs {
  margin: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.currentSiteTab,
.otherSiteTab {
  @include typography.font-size(-1);
  align-items: center;
  appearance: none;
  background: none;
  border: 0;
  color: colours.$site-distinction-tab-text;
  display: flex;
  flex-direction: row;
  font-family: Arial, Helvetica, sans-serif;
  font-weight: 700;
  gap: utils.rem(spacing.$x-small);
  justify-content: center;
  letter-spacing: 1px;
  line-height: 1.5rem;
  padding: utils.rem(spacing.$small) utils.rem(spacing.$small)
    calc(utils.rem(spacing.$small) + 1px); // Extra 1px ensures that the tab aligns perfectly with the sticky nav
  position: relative;
  text-decoration: none;
  text-shadow: 1px 0 0 currentColor;
  top: -1px; // To avoid a 1px gap in Chrome

  &:hover {
    color: colours.$site-distinction-tab-text;
  }
}

.currentSiteTab {
  z-index: 1; // To sit above the other site tab

  @include utils.when-is(button) {
    cursor: pointer;
  }

  &:focus {
    @include focus.default-focus-style;
  }

  &:focus:not(:focus-visible) {
    outline: none;
  }

  &:focus-visible {
    @include focus.default-focus-style;
  }

  .bnf & {
    background: settings.$bnf-brand-colour;
  }

  .bnfc & {
    background: settings.$bnfc-brand-colour;
  }
}

.otherSiteTab {
  cursor: pointer;
  opacity: 0;
  transform: translateY(-100%);
  transition: all 0.25s ease;

  &:hover {
    text-decoration: underline;
    text-decoration-thickness: utils.rem(spacing.$xx-small);
  }

  &:focus {
    z-index: 1; // So focus ring sits above current site tab
  }

  &:visited {
    color: colours.$site-distinction-tab-text;
  }

  .bnf & {
    background: settings.$bnfc-brand-colour;
  }

  .bnfc & {
    background: settings.$bnf-brand-colour;
  }

  &.otherSiteTabExpanded {
    opacity: 1;
    transform: translateY(0);
    z-index: 1; // So tab appears on IE11
  }
}

.icon {
  position: relative;
  top: 0;
  transition: all 0.25s ease;

  .currentSiteTab:hover & {
    transform: translateY(#{utils.rem(spacing.$xx-small)});

    &.iconExpanded {
      transform: translateY(-#{utils.rem(spacing.$xx-small)}) rotate(180deg);
    }
  }

  &.iconExpanded {
    transform: rotate(180deg);
  }
}

// temp width adjustment so tabs don't overlap breadcrumb on mobile
:global(.breadcrumbs) {
  width: 80%;
}

@include media-queries.mq($from: md) {
  :global(.breadcrumbs) {
    width: 100%;
  }
}
