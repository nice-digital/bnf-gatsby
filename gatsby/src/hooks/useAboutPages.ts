import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface AboutPages {
	allBnfAboutSection: {
		aboutSectionPages: SlugAndTitle[];
	};
	allBnfCautionaryAndAdvisoryGuidance: {
		labelsGuidancePages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/about/${slug}/`,
});

/**
 * Gets the lists of pages for the about section by combining:
 * - about sections
 * - cautionary/advisory labels
 * - cautionary/advisory guidance page(s)
 * from the feed.
 *
 * @returns The consolidated list of about pages
 */
export const useAboutPages = (): MenuPageLink[] => {
	const { allBnfAboutSection, allBnfCautionaryAndAdvisoryGuidance } =
		useStaticQuery<AboutPages>(
			graphql`
				query AboutPages {
					allBnfAboutSection(sort: { order: ASC }) {
						aboutSectionPages: nodes {
							slug
							title
						}
					}
					allBnfCautionaryAndAdvisoryGuidance {
						labelsGuidancePages: nodes {
							title
							slug
						}
					}
				}
			`
		);

	return [
		...allBnfAboutSection.aboutSectionPages.map(slugToHref),
		{ title: "Labels", href: "/about/labels/" },
		...allBnfCautionaryAndAdvisoryGuidance.labelsGuidancePages.map(slugToHref),
	];
};
