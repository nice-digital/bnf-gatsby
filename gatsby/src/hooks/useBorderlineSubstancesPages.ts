import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
