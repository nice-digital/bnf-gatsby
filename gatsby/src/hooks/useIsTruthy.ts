import { useCallback } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useIsTruthy = () =>
	useCallback(
		<TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe => !!maybeT,
		[]
	);
