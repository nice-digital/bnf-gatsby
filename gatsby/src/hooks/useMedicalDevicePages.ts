import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicalDevicesQueryResult {
	allBnfMedicalDevice: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medical-devices/${slug}/`,
});

/**
 * Gets the lists of pages for the medical devices section
 *
 * @returns The consolidated list of medical devices pages
 */
export const useMedicalDevicePages = (): MenuPageLink[] => {
	const { allBnfMedicalDevice } = useStaticQuery<MedicalDevicesQueryResult>(
		graphql`
			{
				allBnfMedicalDevice(sort: { fields: title, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfMedicalDevice.pages.map(slugToHref);
};
