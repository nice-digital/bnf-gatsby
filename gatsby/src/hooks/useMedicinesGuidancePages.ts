import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicinesGuidanceQueryResult {
	allBnfGuidance: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medicines-guidance/${slug}/`,
});

/**
 * Gets the lists of pages for the medicines guidance section
 *
 * @returns The consolidated list of medicines guidance pages
 */
export const useMedicinesGuidancePages = (): MenuPageLink[] => {
	const { allBnfGuidance } = useStaticQuery<MedicinesGuidanceQueryResult>(
		graphql`
			{
				allBnfGuidance(sort: { fields: order, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfGuidance.pages.map(slugToHref);
};
