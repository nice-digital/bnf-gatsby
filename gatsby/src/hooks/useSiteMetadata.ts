import { useStaticQuery, graphql } from "gatsby";

interface SiteMetaData {
	/** Whether the current site is BNF or BNFC */
	isBNF: boolean;
	/** Absolute URL to the site homepage, without trailing slash */
	siteUrl: string;
	/** Short site title e.g. "BNF" or "BNFC" */
	siteTitleShort: string;
	/** Full site title e.g. "British National Formulary" or "British National Formulary for Children" */
	siteTitleLong: string;
}

interface Site {
	site: {
		siteMetadata: SiteMetaData;
	};
}

/**
 * Gets the site metadata, as defined in gatsby-config
 *
 * @returns The site metadata
 */
export const useSiteMetadata = (): SiteMetaData => {
	const { site } = useStaticQuery<Site>(
		graphql`
			query SiteMetaData {
				site {
					siteMetadata {
						isBNF
						siteUrl
						siteTitleShort
						siteTitleLong
					}
				}
			}
		`
	);
	return site.siteMetadata;
};
