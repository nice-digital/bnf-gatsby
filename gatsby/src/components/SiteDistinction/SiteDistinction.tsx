import { useLocation } from "@reach/router";
import classNames from "classnames";
import { useMemo, useState, type FC } from "react";

import ChevronDownIcon from "@nice-digital/icons/lib/ChevronDown";
import ChevronRightIcon from "@nice-digital/icons/lib/ChevronRight";
import { Container } from "@nice-digital/nds-container";

import { useIsClient } from "@/hooks/useIsClient";
import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import styles from "./SiteDistinction.module.scss";

export const SiteDistinction: FC = () => {
	const { href } = useLocation(),
		isClient = useIsClient(),
		{ isBNF, siteTitleShort } = useSiteMetadata(),
		[isExpanded, setIsExpanded] = useState(false),
		otherSiteHref = useMemo(() => {
			const url = new URL(href);
			url.host = url.host.replace(
				isBNF ? "bnf" : "bnfc",
				isBNF ? "bnfc" : "bnf"
			);
			url.searchParams.append("ref", "switch");
			return url.toString();
		}, [href, isBNF]);

	return (
		<div
			className={classNames(styles.wrapper, isBNF ? styles.bnf : styles.bnfc)}
		>
			<Container className={styles.container}>
				<p className={styles.tabs}>
					<span className="visually-hidden">
						Currently viewing {siteTitleShort}.
					</span>
					{isClient ? (
						<button
							type="button"
							className={classNames(
								styles.currentSiteTab,
								styles[siteTitleShort]
							)}
							aria-label={`${isExpanded ? "Hide" : "Show"} ${
								isBNF ? "BNFC" : "BNF"
							} link`}
							aria-expanded={isExpanded}
							data-tracking={`${isExpanded ? "hide" : "show"}-${
								isBNF ? "bnfc" : "bnf"
							}-link`}
							onClick={() => setIsExpanded((isExpanded) => !isExpanded)}
						>
							<ChevronDownIcon
								className={classNames(
									styles.icon,
									isExpanded && styles.iconExpanded
								)}
							/>
							{siteTitleShort}
						</button>
					) : (
						<span className={styles.currentSiteTab}>{siteTitleShort}</span>
					)}
					<a
						className={classNames(
							styles.otherSiteTab,
							(isExpanded || !isClient) && styles.otherSiteTabExpanded
						)}
						href={otherSiteHref}
						data-tracking={`${isBNF ? "bnfc" : "bnf"}-link`}
						tabIndex={isExpanded || !isClient ? undefined : -1}
					>
						<span className="visually-hidden">switch to </span>
						{isBNF ? "BNFC" : "BNF"}
						{/* The right chevron is our non-JS fallback to make the other site tab look clickable */}
						{isClient ? null : <ChevronRightIcon className={styles.icon} />}
					</a>
				</p>
			</Container>
		</div>
	);
};
