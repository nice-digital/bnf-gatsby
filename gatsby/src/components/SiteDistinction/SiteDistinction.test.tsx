import { fireEvent, render, screen, within } from "@testing-library/react";

import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import { SiteDistinction } from "./SiteDistinction";

describe("SiteDistinction", () => {
	it("should match snapshot", () => {
		render(<SiteDistinction />);
		expect(screen.getByLabelText("Show BNFC link")).toMatchSnapshot();
	});

	describe.each([
		["BNF", true, "BNFC"],
		["BNFC", false, "BNF"],
	])("%s site", (siteTitleShort, isBNF, otherSiteTitleShort) => {
		beforeEach(() => {
			(useSiteMetadata as jest.Mock).mockReturnValue({
				siteTitleShort,
				isBNF,
			});
		});

		describe("Button", () => {
			it.todo("should not render a button on serverside");
			it(`should render a button with an accessible label to show ${otherSiteTitleShort} link`, () => {
				render(<SiteDistinction />);
				expect(
					screen.getByRole("button", {
						name: `Show ${otherSiteTitleShort} link`,
					})
				).toBeInTheDocument();
			});

			it("should not be expanded by default", () => {
				render(<SiteDistinction />);
				expect(
					screen.getByRole("button", {
						name: `Show ${otherSiteTitleShort} link`,
					})
				).toHaveAttribute("aria-expanded", "false");
			});

			it("should expand on click", () => {
				render(<SiteDistinction />);
				const button = screen.getByRole("button", {
					name: `Show ${otherSiteTitleShort} link`,
				});
				fireEvent.click(button);
				expect(button).toHaveAttribute("aria-expanded", "true");
			});

			it.todo("should change text to hide other site link on click");
			it.todo("should render an icon with an expanded class");
		});

		describe("Link", () => {
			it.todo("should ");
			it.todo("should have an href with ref = switch on querystring");
		});

		describe("Visually hidden", () => {
			it(`should render visually hidden content for ${otherSiteTitleShort} site switch`, () => {
				render(<SiteDistinction />);

				const hiddenSpan = within(
					screen.getByRole("link", { name: `switch to ${otherSiteTitleShort}` })
				).getByText("switch to");
				expect(hiddenSpan).toHaveClass("visually-hidden");
			});
		});
	});
});
