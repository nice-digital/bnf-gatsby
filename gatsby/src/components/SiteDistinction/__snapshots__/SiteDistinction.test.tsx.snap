// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteDistinction BNF site should match snapshot 1`] = `
<button
  aria-expanded="false"
  aria-label="Show BNFC link"
  class="currentSiteTab BNF"
  data-tracking="show-bnfc-link"
  type="button"
>
  <svg
    aria-hidden="true"
    class="icon"
    fill="none"
    focusable="false"
    height="1em"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
      fill="currentColor"
    />
  </svg>
  BNF
</button>
`;

exports[`SiteDistinction BNFC site should match snapshot 1`] = `
<button
  aria-expanded="false"
  aria-label="Show BNF link"
  class="currentSiteTab BNFC"
  data-tracking="show-bnf-link"
  type="button"
>
  <svg
    aria-hidden="true"
    class="icon"
    fill="none"
    focusable="false"
    height="1em"
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
      fill="currentColor"
    />
  </svg>
  BNFC
</button>
`;
